### PR TITLE
build: update k8s version to 1.26: Electrifying

### PR DESCRIPTION
**Pull request by parth-gr**
**Description of your changes:**

build: update k8s version to 1.26: Electrifying

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by subhamkrai — CHANGES_REQUESTED** [review]
Please fix the CI issues

**Review by subhamkrai — CHANGES_REQUESTED** [review]
So I spent some time on this, we need to upgrade the controller-gen version which is used to generate the crds here https://github.com/rook/rook/blob/fcc2786ad0a1c673ae8ca0051a0917b6a9d8e07a/Makefile#L35

also, we need remove an args `,trivialVersions=true` from here https://github.com/rook/rook/blob/fcc2786ad0a1c673ae8ca0051a0917b6a9d8e07a/build/crds/build-crds.sh#L28 it is not working with latest controllergen version.

there doc says it works https://book.kubebuilder.io/reference/controller-gen.html
```controller-gen paths=./... crd:trivialVersions=true rbac:roleName=controller-perms output:crd:artifacts:config=config/crd/bases
```

but cli is not showing that 
```yaml
+crd[:allowDangerousTypes=<bool>][,crdVersions=<[]string>][,generateEmbeddedObjectMeta=<bool>][,ignoreUnexportedFields=<bool>][,maxDescLen=<int>]  package  generates CustomResourceDefinition objects.```

**Comment by subhamkrai** [comment]
> So I spent some time on this, we need to upgrade the controller-gen version which is used to generate the crds here
> 
> https://github.com/rook/rook/blob/fcc2786ad0a1c673ae8ca0051a0917b6a9d8e07a/Makefile#L35
> 
> also, we need remove an args `,trivialVersions=true` from here
> 
> https://github.com/rook/rook/blob/fcc2786ad0a1c673ae8ca0051a0917b6a9d8e07a/build/crds/build-crds.sh#L28
> 
> it is not working with latest controllergen version.
> there doc says it works https://book.kubebuilder.io/reference/controller-gen.html
> 
> but cli is not showing that
> 
> ```yaml
> +crd[:allowDangerousTypes=<bool>][,crdVersions=<[]string>][,generateEmbeddedObjectMeta=<bool>][,ignoreUnexportedFields=<bool>][,maxDescLen=<int>]  package  generates CustomResourceDefinition objects.```
> ```

@travisn and @BlaineEXE  to confirm

**Comment by parth-gr** [comment]
Hey @subhamkrai 
I did the above changes but still, I am not able to create the CRDS, move controller-gen v0.11.3 version,

**Comment by subhamkrai** [comment]
> Hey @subhamkrai I did the above changes but still, I am not able to create the CRDS, move controller-gen v0.11.3 version,

for me locally, it is working with upgrading controller-gen 😅 , I'm not able to report this locally

```yaml
 crds.yaml 
customresourcedefinition.apiextensions.k8s.io/cephblockpoolradosnamespaces.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephblockpools.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephbucketnotifications.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephbuckettopics.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephclients.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephclusters.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephfilesystemmirrors.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephfilesystems.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephfilesystemsubvolumegroups.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephnfses.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectrealms.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectstores.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectstoreusers.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectzonegroups.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectzones.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephrbdmirrors.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/objectbucketclaims.objectbucket.io created
customresourcedefinition.apiextensions.k8s.io/objectbuckets.objectbucket.io created
~/go/src/github.com/rook/deploy/examples
srai@192 ~ (fix-csv-description) $ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1", GitCommit:"8f94681cd294aa8cfd3407b8191f6c70214973a4", GitTreeState:"clean", BuildDate:"2023-01-18T15:58:16Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1", GitCommit:"8f94681cd294aa8cfd3407b8191f6c70214973a4", GitTreeState:"clean", BuildDate:"2023-01-18T15:51:25Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Review by subhamkrai — CHANGES_REQUESTED** [review]
remove every change related to upgrading controller-gen version. And, somewhere in `build-crds.sh` file or any right  place put `sed -i '/x-kubernetes-list-type: atomic/d' crds.yaml` or `sed -i '/x-kubernetes-list-type: atomic/d' resource.yaml` this worked in CI too.


```yaml
runner@fv-az353-302:~/work/rook/rook/deploy/examples$   kubectl delete -f crds.yaml                                                                                                                                                          
 
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephblockpoolradosnamespaces.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephblockpools.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephbucketnotifications.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephbuckettopics.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephclients.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephclusters.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephfilesystemmirrors.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephfilesystems.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephfilesystemsubvolumegroups.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephnfses.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephobjectrealms.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephobjectstores.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephobjectstoreusers.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephobjectzonegroups.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephobjectzones.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "cephrbdmirrors.ceph.rook.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "objectbucketclaims.objectbucket.io" not found
Error from server (NotFound): error when deleting "crds.yaml": customresourcedefinitions.apiextensions.k8s.io "objectbuckets.objectbucket.io" not found
runner@fv-az353-302:~/work/rook/rook/deploy/examples$   kubectl create -f crds.yaml                                                                                                                                                          
 
customresourcedefinition.apiextensions.k8s.io/cephblockpoolradosnamespaces.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephblockpools.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephbucketnotifications.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephbuckettopics.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephclients.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephclusters.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephfilesystemmirrors.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephfilesystems.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephfilesystemsubvolumegroups.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephnfses.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectrealms.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectstores.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectstoreusers.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectzonegroups.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephobjectzones.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/cephrbdmirrors.ceph.rook.io created
customresourcedefinition.apiextensions.k8s.io/objectbucketclaims.objectbucket.io created
customresourcedefinition.apiextensions.k8s.io/objectbuckets.objectbucket.io created
```

**Comment by parth-gr** [comment]
> sed -i '/x-kubernetes-list-type: atomic/d' resource.yaml

Yaa it would work, But I don't so removing them would be a right soution

**Comment by parth-gr** [comment]
The failure is coming from this new validation added 
CustomResourceValidationExpressions https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules
Which is by default on,
I am not sure how we can turn it off in kubelet.

File: Makefile [hunk 1]
```diff
@@ -32,7 +32,7 @@ all: build
 
 # Controller-gen version
 # f284e2e8... is master ahead of v0.5.0 which has ability to generate embedded objectmeta in CRDs
-CONTROLLER_GEN_VERSION=v0.6.2
+CONTROLLER_GEN_VERSION=v0.11.3
```
Comment (subhamkrai) [inline Makefile:35]:
```suggestion
CONTROLLER_GEN_VERSION=v0.11.3
```
can we do the controller runtime changes in the different pr?
since this are not related to k8s 1.26, right?

**Review by subhamkrai — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (parth-gr) [reply Makefile:35]: Actually they are too dependent,
We need min  v0.11.0 https://github.com/kubernetes-sigs/controller-tools/releases/tag/v0.11.0 to support it

**Comment by subhamkrai** [comment]
> Upgrade test is failing
> 
> ```
> Error:      	Received unexpected error:
>         	            	Could Not create resource in args : [apply -f -] -- Failed to run stdin: kubectl [apply -f -] : Warning: resource customresourcedefinitions/cephblockpoolradosnamespaces.ceph.rook.io is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically
> ```
> 
> Any quick suggestions?

```
The CustomResourceDefinition "cephnfses.ceph.rook.io" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
``` this is the real issue 😅

**Comment by parth-gr** [comment]
> > Upgrade test is failing
> > ```
> > Error:      	Received unexpected error:
> >         	            	Could Not create resource in args : [apply -f -] -- Failed to run stdin: kubectl [apply -f -] : Warning: resource customresourcedefinitions/cephblockpoolradosnamespaces.ceph.rook.io is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any quick suggestions?
> 
> ```
> The CustomResourceDefinition "cephnfses.ceph.rook.io" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
> ```
> 
> this is the real issue sweat_smile

Got it thanks,
Used `--server-side=true` for that

File: tests/framework/utils/k8s_helper.go [hunk 2]
```diff
@@ -227,6 +227,9 @@ func (k8sh *K8sHelper) ExecToolboxWithRetry(retries int, namespace, command stri
 // ResourceOperation performs a kubectl action on a pod definition
 func (k8sh *K8sHelper) ResourceOperation(action string, manifest string) error {
 	args := []string{action, "-f", "-"}
+	if action == "apply" {
+		args = []string{"apply", "--server-side=true", "-force-conflicts", "-f", "-"}
```
Comment (BlaineEXE) [inline tests/framework/utils/k8s_helper.go:231]: Is this extra `-` I'm suggesting necessary? Or does kubectl accept single `-` flags?

```suggestion
		args = []string{"apply", "--server-side=true", "--force-conflicts", "-f", "-"}
```

Comment (parth-gr) [reply tests/framework/utils/k8s_helper.go]: THanks @BlaineEXE https://github.com/rook/rook/actions/runs/4244696812/jobs/7379183150

THe upgrade test is in good shape now:)

**Comment by BlaineEXE** [comment]
Closing in favor of https://github.com/rook/rook/pull/11740
